src/runtime_src/core/tools/xbutil2/OO_MemWrite.cpp

```diff
@@ -104,49 +115,145 @@ XBU::verbose("SubCommand option: read mem");
+    std::ifstream input_stream(m_inputFile, std::ios::binary);
+    // If count is unspecified, calculate based on the file size and block size
+    uint64_t count = m_count;
+    if (vm["count"].defaulted()) {
+      input_stream.seekg(0, input_stream.end);
+      // tellg returns a signed value as the number of bytes. Validate the return code and
+      // cast it into a useful format
+      auto nonvalidated_length = input_stream.tellg();
+      if (nonvalidated_length < 0)
+        throw std::runtime_error("Failed to get input file length");
+      uint64_t validated_length = static_cast<uint64_t>(nonvalidated_length);
+
+      if (m_sizeBytes.empty()) // update size
+        size = validated_length;
+      // Set count such that the entire input stream is written to the device
+      // Add size to the validated length to account for truncation due to division
+      count = (validated_length + size - 1) / size;
+      input_stream.seekg(0, input_stream.beg);
+    }
+    else
+      count = m_count;
+
+    if (count <= 0)
+      throw xrt_core::error(std::errc::operation_canceled, "Value for --count must be greater than 0");
+
+    // Logging information
+    XBU::verbose(boost::str(boost::format("Device: %s") % xrt_core::query::pcie_bdf::to_string(xrt_core::device_query<xrt_core::query::pcie_bdf>(device))));
+    XBU::verbose(boost::str(boost::format("Address: %s") % addr));
+    XBU::verbose(boost::str(boost::format("Size: %llu") % size));
+    XBU::verbose(boost::str(boost::format("Block count: %llu") % count));
+    XBU::verbose(boost::str(boost::format("Input File: %s") % m_inputFile));
+    XBU::verbose(boost::str(boost::format("Bytes to write: %lld") % 5));
+
+    // Write to device memory
+    XBU::xclbin_lock xclbin_lock(device);
+    
+    try {
+      for (decltype(count) c = 0; c < count; c++) {
+        XBU::verbose(boost::str(boost::format("[%d / %llu] Writing to Address: %s, Size: %llu bytes") % c % count % addr % size));
+        std::vector<char> buffer(size);
+        // Populate the buffer with the data to write and get the number of bytes read
+        // gcount will only return a value >= 0
+        auto input_size = static_cast<uint64_t>(input_stream.read(buffer.data(), size).gcount());
+        xrt_core::device_mem_write(device.get(), addr, buffer);
+        if (input_size != size)
+          break; // partial read and break the loop
+        addr += size;
+      }
+    } catch (const xrt_core::error& e) {
+      std::cerr << e.what() << std::endl;
+      return;
+    }
+    std::cout << "Memory write succeeded" << std::endl;
+
+    return;
+  }
+
+  // Parse the fill option path
   if(!m_fill.empty()) {
+    // Validate the block count. This cannot be done as the --input path may create its own block count
+    if (m_count <= 0)
+      throw xrt_core::error(std::errc::operation_canceled, "Value for --count must be greater than 0");
+    
+    // Validate the number of bytes to write
+    if (m_sizeBytes.empty())
```

Comment:
What happens if the user specifies a size of 0? :^)

Reply:
In this case the code will continue and eventually write nothing